### PR TITLE
Update card message using ReplyToid

### PR DESCRIPTION
**Pull request by vamagra**
@msftdherron @jotrick @MattSFT @aosolis @goelashish7 @jialic @robin-liao

### CODE REVIEW

File: src/dialogs/examples/teams/UpdateCardMsgSetupDialog.ts [hunk 1]
```diff
@@ -3,23 +3,38 @@ import { TriggerActionDialog } from "../../../utils/TriggerActionDialog";
...
+        let messageBackButtonValue = JSON.stringify({ updateCounterKey: ++updateCardCounter });
+        let messageBackButton = builder.CardAction.messageBack(session, messageBackButtonValue)
+            .displayText(Strings.messageBack_button_display_text)
+            .title(Strings.update_card_button, updateCardCounter)
+            .text("update card message");
+
+        buttons.push(messageBackButton);
+
+        let card = new builder.HeroCard(session)
+        .title(Strings.default_title)
```
Comment (aosolis) [inline src/dialogs/examples/teams/UpdateCardMsgSetupDialog.ts]: Nit: consistently indent these lines

File: src/dialogs/examples/teams/UpdateCardMsgSetupDialog.ts [hunk 2]
```diff
@@ -3,23 +3,38 @@ import { TriggerActionDialog } from "../../../utils/TriggerActionDialog";
...
+let config = require("config");
 
 export class UpdateCardMsgSetupDialog extends TriggerActionDialog {
-
+    // setup the card message and then user can update the card using below update dialog file
+    // microsoft-teams-sample-complete-node\src\dialogs\examples\teams\UpdateCardMsgDialog.ts
     private static async setupCardMessage(session: builder.Session, args?: any | builder.IDialogResult<any>, next?: (args?: builder.IDialogResult<any>) => void): Promise<void> {
-        let tempCard = new builder.HeroCard(session).title(Strings.loading);
-        let msg = new builder.Message(session).addAttachment(tempCard);
+        let buttons = new Array<builder.CardAction>();
```
Comment (aosolis) [inline src/dialogs/examples/teams/UpdateCardMsgSetupDialog.ts]: There's only 1 button, no need for this array, just do .buttons([messageBackButton])

File: src/dialogs/examples/teams/UpdateCardMsgSetupDialog.ts [hunk 3]
```diff
@@ -3,23 +3,38 @@ import { TriggerActionDialog } from "../../../utils/TriggerActionDialog";
 import { DialogIds } from "../../../utils/DialogIds";
 import { DialogMatches } from "../../../utils/DialogMatches";
 import { Strings } from "../../../locale/locale";
+let config = require("config");
```
Comment (aosolis) [inline src/dialogs/examples/teams/UpdateCardMsgSetupDialog.ts]: I think we normally do 
import * as config from "config"

File: src/dialogs/examples/teams/UpdateCardMsgDialog.ts [hunk 4]
```diff
@@ -6,49 +6,49 @@ import { Strings } from "../../../locale/locale";
...
-                .type("invoke")
-                .title(Strings.update_card_button, session.conversationData.updateCardCounter)
-                .value("{" +
-                    "\"dialog\": \"" + DialogIds.UpdateCardMsgDialogId + "\", " +
-                    "\"address\": " + JSON.stringify(args.address) + "" +
-                "}"),
-            );
+            let messageBackButton = builder.CardAction.messageBack(session, messageBackButtonValue)
+                .displayText(Strings.messageBack_button_display_text)
+                .title(Strings.update_card_button, updateCardCounter)
```
Comment (aosolis) [inline src/dialogs/examples/teams/UpdateCardMsgDialog.ts]: Can we keep this as always "Update card", and instead change the card title to be "Update #{updateCount}"?

File: src/dialogs/examples/teams/UpdateCardMsgSetupDialog.ts (hunk 2 above)
Comment (aosolis) [reply src/dialogs/examples/teams/UpdateCardMsgSetupDialog.ts]: Same above in UpdateCardMsgDialog

File: src/dialogs/examples/teams/UpdateCardMsgDialog.ts [hunk 5]
```diff
@@ -6,49 +6,49 @@ import { Strings } from "../../../locale/locale";
...
                 .text(Strings.default_text)
                 .images([
                     new builder.CardImage(session)
                         .url(config.get("app.baseUri") + "/assets/computer_person.jpg")
                         .alt(session.gettext(Strings.img_default)),
-                ])
+                    ])
                 .buttons(buttons);
 
+            let newAddress = { ...session.message.address, id: session.message.replyToId };
```
Comment (aosolis) [inline src/dialogs/examples/teams/UpdateCardMsgDialog.ts]: newAddress is misleading-- this is actually the address of the *existing* message, so it's not at all new.
I'd change to addressOfMessageToUpdate

File: src/dialogs/examples/teams/UpdateCardMsgDialog.ts [hunk 6]
```diff
@@ -59,7 +59,7 @@ export class UpdateCardMsgDialog extends TriggerActionDialog {
     ) {
         super(bot,
             DialogIds.UpdateCardMsgDialogId,
-            DialogMatches.Update_Card_Msg_Dialog_Intent,
+            DialogMatches.UpdateCardMsgDialogMatch,
```
Comment (aosolis) [inline src/dialogs/examples/teams/UpdateCardMsgDialog.ts]: Please change back to Update_Card_Msg_Dialog_Intent, and use that constant as the text of the messageback button.

File: src/dialogs/examples/teams/UpdateCardMsgDialog.ts (hunk 4 above)
Comment (vamagra) [reply src/dialogs/examples/teams/UpdateCardMsgDialog.ts]: okay, then I have to do the same changed in C# sample as well to keep the consistency.

File: src/dialogs/examples/teams/UpdateCardMsgSetupDialog.ts (hunk 1 above)
Comment (vamagra) [reply src/dialogs/examples/teams/UpdateCardMsgSetupDialog.ts]: Done.

File: src/dialogs/examples/teams/UpdateCardMsgSetupDialog.ts (hunk 2 above)
Comment (vamagra) [reply src/dialogs/examples/teams/UpdateCardMsgSetupDialog.ts]: Done.

File: src/dialogs/examples/teams/UpdateCardMsgSetupDialog.ts (hunk 3 above)
Comment (vamagra) [reply src/dialogs/examples/teams/UpdateCardMsgSetupDialog.ts]: Done.

File: src/dialogs/examples/teams/UpdateCardMsgDialog.ts (hunk 5 above)
Comment (vamagra) [reply src/dialogs/examples/teams/UpdateCardMsgDialog.ts]: Done.

File: src/dialogs/examples/teams/UpdateCardMsgDialog.ts (hunk 6 above)
Comment (vamagra) [reply src/dialogs/examples/teams/UpdateCardMsgDialog.ts]: Done.

Comment (aosolis) [reply src/dialogs/examples/teams/UpdateCardMsgDialog.ts]: No, don't change Update_Card_Msg_Dialog_Intent to a regexp. This is only used to route the message; the user should never type the text herself, so there's no need for a regexp. Then you can use the string constant for the text of the messageBack action.

File: src/dialogs/examples/teams/UpdateCardMsgSetupDialog.ts [hunk 7]
```diff
@@ -3,23 +3,35 @@ import { TriggerActionDialog } from "../../../utils/TriggerActionDialog";
...
     private static async setupCardMessage(session: builder.Session, args?: any | builder.IDialogResult<any>, next?: (args?: builder.IDialogResult<any>) => void): Promise<void> {
-        let tempCard = new builder.HeroCard(session).title(Strings.loading);
-        let msg = new builder.Message(session).addAttachment(tempCard);
+        let updateCardCounter = 0;
+
+        let messageBackButtonValue = JSON.stringify({ updateCounterKey: updateCardCounter });
+        let messageBackButton = builder.CardAction.messageBack(session, messageBackButtonValue)
+            .displayText(Strings.messageBack_button_display_text)
+            .title(Strings.update_card_button)
+            .text("update card message");
```
Comment (aosolis) [inline src/dialogs/examples/teams/UpdateCardMsgSetupDialog.ts]: Add a comment here saying
// This must be a string that routes to UpdateCardMsgDialog, which handles card updates

Also in UpdateCardMsgDialog itself at line 20 (where you set the text of the messageBack action)

Comment (vamagra) [reply src/dialogs/examples/teams/UpdateCardMsgSetupDialog.ts]: Done.

**Pull request merged**